homeassistant/components/iqvia/sensor.py

```diff
@@ -0,0 +1,210 @@
+"""Support for IQVIA sensors."""
+import logging
+from statistics import mean
+
+from homeassistant.components.iqvia import (
+    DATA_CLIENT, DOMAIN, SENSORS, TYPE_ALLERGY_FORECAST, TYPE_ALLERGY_OUTLOOK,
+    TYPE_ALLERGY_INDEX, TYPE_ALLERGY_TODAY, TYPE_ALLERGY_TOMORROW,
+    TYPE_ALLERGY_YESTERDAY, TYPE_ASTHMA_INDEX, TYPE_ASTHMA_TODAY,
+    TYPE_ASTHMA_TOMORROW, TYPE_ASTHMA_YESTERDAY, IQVIAEntity)
+from homeassistant.const import ATTR_STATE
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_ALLERGEN_AMOUNT = 'allergen_amount'
+ATTR_ALLERGEN_GENUS = 'allergen_genus'
+ATTR_ALLERGEN_NAME = 'allergen_name'
+ATTR_ALLERGEN_TYPE = 'allergen_type'
+ATTR_CITY = 'city'
+ATTR_OUTLOOK = 'outlook'
+ATTR_RATING = 'rating'
+ATTR_SEASON = 'season'
+ATTR_TREND = 'trend'
+ATTR_ZIP_CODE = 'zip_code'
+
+RATING_MAPPING = [{
+    'label': 'Low',
+    'minimum': 0.0,
+    'maximum': 2.4
+}, {
+    'label': 'Low/Medium',
+    'minimum': 2.5,
+    'maximum': 4.8
+}, {
+    'label': 'Medium',
+    'minimum': 4.9,
+    'maximum': 7.2
+}, {
+    'label': 'Medium/High',
+    'minimum': 7.3,
+    'maximum': 9.6
+}, {
+    'label': 'High',
+    'minimum': 9.7,
+    'maximum': 12
+}]
+
+TREND_INCREASING = 'Increasing'
+TREND_SUBSIDING = 'Subsiding'
+
+
+async def async_setup_platform(
+        hass, config, async_add_entities, discovery_info=None):
+    """Configure the platform and add the sensors."""
+    iqvia = hass.data[DOMAIN][DATA_CLIENT]
+
+    sensors = []
+    for kind in iqvia.sensor_types:
+        sensor_class, name, icon = SENSORS[kind]
+        sensors.append(
+            globals()[sensor_class](iqvia, kind, name, icon, iqvia.zip_code))
+
+    async_add_entities(sensors, True)
+
+
+def calculate_average_rating(indices):
+    """Calculate the human-friendly historical allergy average."""
+    ratings = list(
+        r['label'] for n in indices for r in RATING_MAPPING
+        if r['minimum'] <= n <= r['maximum'])
+    return max(set(ratings), key=ratings.count)
+
+
+def calculate_trend(indices):
+    """Calculate the "moving average" of a set of indices."""
+    import numpy as np
+
+    def moving_average(data, samples):
+        """Determine the "moving average" (http://tinyurl.com/yaereb3c)."""
+        ret = np.cumsum(data, dtype=float)
+        ret[samples:] = ret[samples:] - ret[:-samples]
+        return ret[samples - 1:] / samples
+
+    increasing = np.all(np.diff(moving_average(np.array(indices), 4)) > 0)
+
+    if increasing:
+        return TREND_INCREASING
+    return TREND_SUBSIDING
+
+
+class ForecastSensor(IQVIAEntity):
+    """Define sensor related to forecast data."""
+
+    async def async_update(self):
+        """Update the sensor."""
+        await self._iqvia.async_update()
```

Comment:
Isn't this called already and that's why we're updating here in the first place? 🥚 🐥